arch/x86/core/spec_ctrl.c

```diff
@@ -0,0 +1,55 @@
+/*
+ * Copyright (c) 2018 Intel Corporation
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <cpuid.h> /* Header provided by the toolchain. */
+
+#include <init.h>
+#include <kernel_structs.h>
+#include <kernel_arch_data.h>
+#include <kernel_arch_func.h>
+#include <kernel.h>
+
+#define CPUID_EXTENDED_FEATURES_LVL 7
+
+#define CPUID_SPEC_CTRL BIT(26)
+#define SPEC_CTRL_SSBD BIT(2)
+
+static int
+cpu_has_spec_ctrl(void)
+{
+	u32_t eax, ebx, ecx = 0, edx;
+
+	if (!__get_cpuid(CPUID_EXTENDED_FEATURES_LVL,
+			 &eax, &ebx, &ecx, &edx)) {
+		return 0;
+	}
+
+	ARG_UNUSED(eax);
+	ARG_UNUSED(ebx);
+	ARG_UNUSED(ecx);
+
+	return edx & CPUID_SPEC_CTRL;
+}
+
+static int
+disable_ssbd_if_needed(struct device *dev)
+{
+	/* This is checked in runtime rather than compile time since
+	 * IA32_SPEC_CTRL_MSR might be added in a microcode update.
+	 */
+	if (cpu_has_spec_ctrl()) {
+		u64_t cur = _MsrRead(IA32_SPEC_CTRL_MSR);
+
+		_MsrWrite(IA32_SPEC_CTRL_MSR,
+			  cur | SPEC_CTRL_SSBD);
```

Comment:
Had to look these up too.  We need to fix our implementation (not an issue with this patch).  _MsrRead/Write are implemented as C functions instead of inlines, and are significantly larger than neecessary because of that (because again rdmsr/wrmsr are single instructions).  Should file a bug.  Easy 10-20 bytes of code savings available.

Reply:
Not worried about shaving a few bytes here; small core x86 targets do not speculate execution, so this wouldn't be enabled anyway.

Reply:
In any case, sent a PR (#7765) for this; renamed the functions to conform to the conventions, and made them static inline.  The compiler should be able to generate better code, given that it now has more information of which registers are being used/clobbered.